src/pipeline/src/etl.rs

```diff
@@ -56,24 +114,92 @@ where
                 Transforms::default()
             };
 
+            let mut transformer = T::new(transforms)?;
+            let transforms = transformer.transforms_mut();
+
+            let processors_output_keys = processors.output_keys();
+            let processors_required_keys = processors.required_keys();
+            let processors_required_original_keys = processors.required_original_keys();
+
+            debug!(
+                "processors_required_original_keys: {:?}",
+                processors_required_original_keys
+            );
+            debug!("processors_required_keys: {:?}", processors_required_keys);
+            debug!("processors_output_keys: {:?}", processors_output_keys);
+
+            let transforms_required_keys = transforms.required_keys();
+            let mut tr_keys = Vec::with_capacity(50);
+            for key in transforms_required_keys.iter() {
+                if !processors_output_keys.contains(key)
+                    && !processors_required_original_keys.contains(key)
+                {
+                    tr_keys.push(key.clone());
+                }
+            }
+
+            let mut required_keys = processors_required_original_keys.clone();
+
+            required_keys.append(&mut tr_keys);
+            required_keys.sort();
+
+            debug!("required_keys: {:?}", required_keys);
+
+            // intermediate keys are the keys that all processor and transformer required
+            let ordered_intermediate_keys: Vec<String> =
+                merge(processors_required_keys, transforms_required_keys)
+                    .cloned()
+                    .collect::<HashSet<String>>()
+                    .into_iter()
+                    .sorted()
+                    .collect();
+
+            let mut final_intermediate_keys = Vec::with_capacity(ordered_intermediate_keys.len());
+            let mut intermediate_keys_exclude_original =
+                Vec::with_capacity(ordered_intermediate_keys.len());
+
+            for key_name in ordered_intermediate_keys.iter() {
+                if required_keys.contains(key_name) {
+                    final_intermediate_keys.push(key_name.clone());
+                } else {
+                    intermediate_keys_exclude_original.push(key_name.clone());
+                }
+            }
+
+            final_intermediate_keys.extend(intermediate_keys_exclude_original);
```

Comment:
So this logic just wants to split the ordered_intermediate_keys by checking if it is in `required_keys`? If so, maybe use a group_by and concat is more clear.

Reply:
Ensure that all user-supplied data required by the processor (except for process output data) fields are at the top of the list and keep the fields in order.